README.md
<h1 align="center">
    Swap File
    <br/><br/>
    <a><img src="https://github.com/mxnt10/swapfile/raw/master/common/img_swapfile.png"></a>
</h1>

Program created with the objective of assisting in the configuration of swap file in Distributions such as Slackware.

### Required packages

- python >= 3.6

### Installation instructions

To install "Swap File" , use the following commands:
```sh
$ git clone https://github.com/mxnt10/swapfile.git
$ cd swapfile

$ sudo ./install.sh
```

### GNU General Public License

This repository has scripts that were created to be free software.<br/>
Therefore, they can be distributed and / or modified within the terms of the *GNU General Public License*.

>[General Public License](https://pt.wikipedia.org/wiki/GNU_General_Public_License)
>
>Free Software Foundation (FSF) Inc. 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA

### Comments

In case of bugs, problems of execution or construction of packages, constructive criticism, among others, please submit a message to one of the contacts below.

### Contact

Maintainer: Mauricio Ferrari<br/>
E-Mail: *m10ferrari1200@gmail.com*<br/>
Telegram: *@maurixnovatrento*<br/>


ChangeLog
# ChangeLog
All notable changes to this project will be documented in this file.
Todas as mudanças notáveis neste projeto serão documentadas neste arquivo.


## [ 1.1 ] - 2021-11-09 #############################################

### Added / Adicionado
 - Added image in common directory.
###

### Fixed / Corrigido
 - Fixed transparency in options.
###

#####################################################################


## [ 1.0 ] - 2021-10-24 #############################################

### Added / Adicionado
    - Add install.sh
    - Add swapfile.png and swapfile.desktop in appdata.
    - Add clean_cache in common.
    - Add sources in src.
    - Add scripts in shell in utils.
    - Include psutil in sources.
###

#####################################################################


src/utils.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

# Import modules
import logging as log
import subprocess as sh
import os

# Import sources
import messagebox as msgbox


# Verify if password is valid
def pass_root(win):
    passwd = None  # It's returned None for root user
    if os.geteuid() != 0:
        loop = 1
        while loop != 0:
            passwd = msgbox.showpass(win, 'Authentication required', 'Root password:')
            if passwd != (-1):
                com = get_local('get_root') + ' ' + passwd
                op = sh.run(com, shell=True)
                exitcode = getattr(op, 'returncode')
                loop = exitcode
                if exitcode != 0:
                    msgbox.showerror(win, 'Error', 'Invalid Password!')
            else:
                loop = 0  # Return 0 if password in None or window as closed.
    return passwd


# Search for files
def get_local(file):
    get_file = '/usr/share/swapfile/utils/' + file
    l_get_file = '../utils/' + file

    try:
        with open(get_file):
            return get_file
    except Exception as msg:
        log.warning("\033[33m %s.\033[32m Use a local icon...\033[m", msg)
        return l_get_file


# Define application icon
def set_icon():
    icon = '/usr/share/pixmaps/swapfile.png'
    l_icon = '../appdata/swapfile.png'

    try:
        with open(icon):
            return icon
    except Exception as msg:
        log.warning("\033[33m %s.\033[32m Use a local icon...\033[m", msg)
        try:
            with open(l_icon):
                return l_icon
        except Exception as msg:
            # Exception for icon not found
            log.error("\033[31m %s \033[m", msg)
            return None


# Open application in screen center
def center(win):
    # :param win: the main window or Toplevel window to center

    # Apparently a common hack to get the window size. Temporarily hide the
    # window to avoid update_idletasks() drawing the window in the wrong
    # position.
    win.update_idletasks()  # Update "requested size" from geometry manager

    # define window dimensions width and height
    width = win.winfo_width()
    frm_width = win.winfo_rootx() - win.winfo_x()
    win_width = width + 2 * frm_width

    height = win.winfo_height()
    titlebar_height = win.winfo_rooty() - win.winfo_y()
    win_height = height + titlebar_height + frm_width

    # Get the window position from the top dynamically as well as position from left or right as follows
    x = win.winfo_screenwidth() // 2 - win_width // 2
    y = win.winfo_screenheight() // 2 - win_height // 2

    # this is the line that will center your window
    win.geometry('{}x{}+{}+{}'.format(width, height, x, y))

    # This seems to draw the window frame immediately, so only call deiconify()
    # after setting correct window position
    win.deiconify()


src/actions.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

# Import modules
from tkinter import Toplevel, Label, Entry, Scale, Frame, Button, IntVar
from tkinter import StringVar as StrVar
import subprocess as sh

# Import sources
import messagebox as msgbox
import calc_mem as mem
import utils as ut


# Active swap in system
def active(win, intvar, st):
    passwd = ut.pass_root(win)
    if passwd != (-1) and passwd is not None:
        act_comm = 'sudo -S -k <<< ' + passwd + ' ' + ut.get_local('active_swap') + ' ' + str(intvar) + 'M'
        sh.run(act_comm, shell=True)
    if passwd is None:
        act_comm = ut.get_local('active_swap') + ' ' + str(intvar) + 'M'
        sh.run(act_comm, shell=True)
    st.config(text='Status: ' + mem.status_swap())


# Deactive swap in system
def deactive(win, st):
    passwd = ut.pass_root(win)
    if passwd != (-1) and passwd is not None:
        act_comm = 'sudo -S -k <<< ' + passwd + ' ' + ut.get_local('deactive_swap')
        sh.run(act_comm, shell=True)
    if passwd is None:
        act_comm = ut.get_local('deactive_swap')
        sh.run(act_comm, shell=True)
    st.config(text='Status: ' + mem.status_swap())


# Define swap priority in system
def apply_swappines(win, intvar):
    passwd = ut.pass_root(win)
    if passwd != (-1) and passwd is not None:
        act_comm = 'sudo -S -k <<< ' + passwd + ' ' + ut.get_local('sysctl_file') + ' ' + str(intvar)
        sh.run(act_comm, shell=True)
    if passwd is None:
        act_comm = ut.get_local('sysctl_file') + ' ' + str(intvar)
        sh.run(act_comm, shell=True)


# Reset swap priority in system
def reset_swappines(win, scale):
    passwd = ut.pass_root(win)
    if passwd != (-1) and passwd is not None:
        act_comm = 'sudo -S -k <<< ' + passwd + ' ' + ut.get_local('sysctl_rst')
        sh.run(act_comm, shell=True)
        scale.config(command='')
        scale.set(mem.get_swappines())
    if passwd is None:
        act_comm = ut.get_local('sysctl_rst')
        sh.run(act_comm, shell=True)
        scale.config(command='')
        scale.set(mem.get_swappines())


# Window for swap options
def options(main):
    # No necessary minimize window
    # noinspection PyUnusedLocal
    def on_unmap(*args):
        op_app.deiconify()
        op_app.focus_force()

    # Active button apply
    # noinspection PyUnusedLocal
    def on_apply(*args):
        b_options['state'] = 'normal'

    # Deactive button apply
    def off_apply():
        b_options['state'] = 'disabled'
        scale_mem.config(command=on_apply)

    op_app = Toplevel()
    op_app.title('Options')
    op_app.resizable(0, 0)

    # Focuses on the window and avoids clicking on the main window
    op_app.transient(main)
    op_app.focus_force()
    op_app.grab_set()

    # Define swappiness value for entry
    swp = StrVar()
    swp.set(mem.get_swappines())

    # Variable for Scale
    i_scale = IntVar()

    # Create frame for swappiness option
    frame = Frame(op_app, borderwidth=2, relief='groove')
    frame.grid(row=0, column=0, padx=(5, 5), pady=(5, 5), sticky='E, W')

    # Create entry for swappiness
    Label(frame, text='Swappiness:').grid(row=0, column=0, padx=(15, 10), pady=(12, 5), sticky='E')
    e_swap = Entry(frame, width=8, justify='right', fg='#000', textvariable=swp)
    e_swap.config(readonlybackground='#fff', state='readonly', highlightthickness=0)
    e_swap.grid(row=0, column=1, padx=(0, 15), pady=(12, 5), sticky='W')

    # Scale for ajust swap option
    scale_mem = Scale(frame, from_=1, to=100, orient='horizontal', length=180, variable=i_scale)
    scale_mem.set(mem.get_swappines())
    scale_mem.grid(row=1, column=0, padx=10, pady=10, columnspan=2)

    # Buttons for options
    b_options = Button(frame, text='Apply', width=8, command=lambda: {apply_swappines(op_app, i_scale.get()),
                                                                      swp.set(mem.get_swappines()),
                                                                      off_apply()})
    b_options.grid(row=2, column=0, padx=2, pady=5)
    off_apply()

    b_reset = Button(frame, text='Reset', width=8, command=lambda: {reset_swappines(op_app, scale_mem),
                                                                    swp.set(mem.get_swappines()),
                                                                    off_apply()})
    b_reset.grid(row=2, column=1, padx=(2, 10), pady=5)

    op_app.bind("<Unmap>", on_unmap)  # For not minimize configuration window
    op_app.wait_visibility(op_app)
    op_app.attributes('-alpha', 0.85)
    op_app.mainloop()


# About application
def about(main):
    msgbox.message(main, 'About', 'SwapFile v1.1 - Swap file manager\n\n'
                                  'Maintainer: Mauricio Ferrari\n'
                                  'Contact: m10ferrari1200@gmail.com\n')


src/calc_mem.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

import math
import psutil
import subprocess as sh
# noinspection PyProtectedMember
from psutil._common import bytes2human


def total_mem():
    mem = psutil.virtual_memory()
    calc = getattr(mem, 'total')
    return (calc / 1024) / 1024


def total_swap():
    swap = psutil.swap_memory()
    return bytes2human(getattr(swap, 'total'))


def used_swap():
    swap = psutil.swap_memory()
    return bytes2human(getattr(swap, 'used'))


def status_swap():
    st = used_swap() + ' / ' + total_swap()
    if total_swap() == '0.0B':
        st += ' - disabled'
    else:
        st += ' - enabled'
    return st


def min_mem():
    if total_mem() < 1024:
        return total_mem()
    else:
        return int(math.sqrt(total_mem() / 1024) * 1024)


def rec_mem():
    if total_mem() < 1024:
        return total_mem() * 2
    else:
        return int(min_mem() + total_mem())


def max_mem():
    return int(total_mem() * 2)


def get_swappines():
    swp = sh.check_output('sysctl -n vm.swappiness', shell=True)
    return str(swp.decode()).strip('\n')


src/messagebox.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

# Import tkinter module
import tkinter as tk
import utils as ut

# Global variable for showpass
passwd = None


# Necessary for no interrupt code
def close_msg(win, strvar=None, close=False):
    win.quit()
    win.destroy()
    global passwd
    if strvar is not None:
        passwd = strvar.get()
    if close:
        passwd = (-1)


# Function for view message
def message(win, title, txt, image='', entry=False):
    # No necessary minimize window

    # noinspection PyUnusedLocal
    def on_unmap(*args):
        msg.deiconify()
        msg.focus_force()

    # noinspection PyUnusedLocal
    def enter_entry(*args):
        close_msg(msg, strvar=strvar)

    # noinspection PyUnusedLocal
    def enter(*args):
        close_msg(msg)

    msg = tk.Toplevel()
    msg.wait_visibility(msg)
    msg.attributes('-alpha', 0.0)
    msg.title(title)
    msg.resizable(False, False)

    # Focuses on the window and avoids clicking on the main window
    msg.transient(win)
    msg.focus_force()
    msg.grab_set()

    # Create frame for full organization
    frame = tk.Frame(msg)
    frame.pack()

    # Label logo
    logo = tk.Label(frame, image=image)
    logo.grid(row=0, column=0, pady=(10, 10), padx=(20, 10))

    # Insert options for entry:
    if entry:
        nm = 0
        strvar = tk.StringVar()
        l_entry = tk.Entry(frame, justify='right', show='*', textvariable=strvar)
        l_entry.grid(row=0, column=2, pady=(10, 10), padx=(10, 30))
        l_entry.focus()
        button = tk.Button(msg, text="OK", command=lambda: close_msg(msg, strvar=strvar), width=10)
        msg.bind('<Return>', enter_entry)
    else:
        nm = 30
        button = tk.Button(msg, text="OK", command=lambda: close_msg(msg), width=10)
        msg.bind('<Return>', enter)

    # Create button
    button.pack(pady=(0, 10))

    # Insert text
    text = tk.Label(frame, text=txt, font='-weight bold -size 10')
    text.grid(row=0, column=1, pady=(10, 10), padx=(10, nm))

    ut.center(msg)
    msg.attributes('-alpha', 1.0)
    msg.bind("<Unmap>", on_unmap)  # For not minimize message dialog
    msg.protocol("WM_DELETE_WINDOW", lambda: close_msg(msg, close=True))
    msg.mainloop()


# Dialog for information
def showpass(win, title, text):
    image = '::tk::icons::information'
    message(win, title, text, image=image, entry=True)
    global passwd
    return passwd


# Dialog for error
def showerror(win, title, text):
    image = '::tk::icons::error'
    message(win, title, text, image=image)


src/swapfile.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-

# Import modules
from tkinter import Tk, Label, Entry, Scale, Frame, Button, IntVar
from tkinter import StringVar as StrVar
from tkinter import PhotoImage as Ico

# Import sources
import actions as ac
import calc_mem as mem
import utils as ut


# Verify scale state
def state_scale(scale):
    if mem.total_swap() == '0.0B':
        scale.config(state='normal', takefocus=1)
        scale.set(mem.min_mem())
    else:
        scale.set(0)
        scale.config(state='disabled', takefocus=0)


# Verify button state
def state_button(act, deact):
    if mem.total_swap() == '0.0B':
        act['state'] = 'normal'
        deact['state'] = 'disabled'
    else:
        act['state'] = 'disabled'
        deact['state'] = 'normal'


# Main function
def launcher_swapfile():
    main_app = Tk()
    main_app.title('Swap File')
    main_app.resizable(0, 0)

    # Set program icon case function is not None
    ico = ut.set_icon()
    if ico is not None:
        main_app.iconphoto(True, Ico(file=ico))

    # Variables for Entry
    s_min = StrVar()
    s_min.set(mem.min_mem())
    s_rec = StrVar()
    s_rec.set(mem.rec_mem())
    s_max = StrVar()
    s_max.set(mem.max_mem())

    # Variable for Scale
    i_scale = IntVar()

    # Functions for select entry e_min
    # noinspection PyUnusedLocal
    def sel_min(*args):
        main_app.focus()
        scale_mem.set(mem.min_mem())

    # Functions for select entry e_min
    # noinspection PyUnusedLocal
    def sel_rec(*args):
        main_app.focus()
        scale_mem.set(mem.rec_mem())

    # Functions for select entry e_min
    # noinspection PyUnusedLocal
    def sel_max(*args):
        main_app.focus()
        scale_mem.set(mem.max_mem())

    # Frame for memory informations
    frame_mem = Frame(main_app, borderwidth=2, relief='groove')
    frame_mem.grid(row=0, column=0, padx=(5, 5), pady=(5, 5), sticky='E, W')

    # Frame for Scale
    frame_scale = Frame(main_app, borderwidth=2, relief='groove')
    frame_scale.grid(row=1, column=0, padx=(5, 5), pady=(0, 5), sticky='E, W')

    # Frame for Buttons
    frame_button = Frame(main_app)
    frame_button.grid(row=2, column=0)

    # For minimal recommended
    Label(frame_mem, text='Minimal recommended:').grid(row=0, column=0, padx=(20, 10), pady=(12, 5), sticky='E')
    Label(frame_mem, text='MB').grid(row=0, column=2, pady=(12, 5), sticky='W')
    e_min = Entry(frame_mem, width=12, justify='right', fg='#000', textvariable=s_min)
    e_min.config(readonlybackground='#fff', state='readonly')
    e_min.bind('<FocusIn>', sel_min)
    e_min.grid(row=0, column=1, pady=(12, 5))

    # For recommended
    Label(frame_mem, text='Recommended:').grid(row=1, column=0, padx=(20, 10), pady=5, sticky='E')
    Label(frame_mem, text='MB').grid(row=1, column=2, pady=5, sticky='W')
    e_rec = Entry(frame_mem, width=12, justify='right', fg='#000', textvariable=s_rec)
    e_rec.config(readonlybackground='#fff', state='readonly')
    e_rec.bind('<FocusIn>', sel_rec)
    e_rec.grid(row=1, column=1, pady=5)

    # For maximum recommended
    Label(frame_mem, text='Maximum recommended:').grid(row=2, column=0, padx=(20, 10), pady=5, sticky='E')
    Label(frame_mem, text='MB').grid(row=2, column=2, padx=(0, 20), pady=5, sticky='W')
    e_max = Entry(frame_mem, width=12, justify='right', fg='#000', textvariable=s_max)
    e_max.config(readonlybackground='#fff', state='readonly')
    e_max.bind('<FocusIn>', sel_max)
    e_max.grid(row=2, column=1, pady=5)

    # Scale for select the amount of RAM memory
    Label(frame_scale, text='Select the amount of RAM memory:').grid(row=0, column=0, padx=10, pady=(12, 5), sticky='W')
    scale_mem = Scale(frame_scale, from_=16, to=mem.max_mem(), orient='horizontal', length=320, variable=i_scale)
    scale_mem.grid(row=1, column=0, padx=10, pady=(0, 10))
    state_scale(scale_mem)

    # Buttons for options
    b_active = Button(frame_button, text='Active', width=8, command=lambda: {ac.active(main_app, i_scale.get(), status),
                                                                             state_button(b_active, b_deactive),
                                                                             state_scale(scale_mem)})
    b_active.grid(row=0, column=0, padx=2, pady=(0, 5))

    b_deactive = Button(frame_button, text='Deactive', width=8, command=lambda: {ac.deactive(main_app, status),
                                                                                 state_button(b_active, b_deactive),
                                                                                 state_scale(scale_mem)})
    b_deactive.grid(row=0, column=1, padx=2, pady=(0, 5))

    b_options = Button(frame_button, text='Options', width=8, command=lambda: ac.options(main_app))
    b_options.grid(row=0, column=2, padx=2, pady=(0, 5))

    b_about = Button(frame_button, text='About', width=8, command=lambda: ac.about(main_app))
    b_about.grid(row=0, column=3, padx=2, pady=(0, 5))

    status = Label(frame_button, text='Status: ' + mem.status_swap())
    status.grid(row=1, column=0, padx=2, pady=5, sticky='W', columnspan=4)

    # Ajust state for Buttons
    state_button(b_active, b_deactive)

    # Others properties
    ut.center(main_app)
    main_app.wait_visibility(main_app)
    main_app.attributes('-alpha', 0.85)
    main_app.mainloop()


# Start application
if __name__ == '__main__':
    launcher_swapfile()
